File: codes/julia/ch21/symbolic_quartic_oscillator.jl
# symbolic_quartic_oscillator.jl
# Chapter 21, Etude 21.10: quartic oscillator on the infinite line via the
# rational-Chebyshev map y = ell x / sqrt(1 - x^2), ell = 2.
# Symbolic 5-term Galerkin produces a 5th-degree secular polynomial in E
# (Boyd Eq 20.16); its roots are compared to Bender-Orszag references.
#
# Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)

using SymPy
using Polynomials
include(joinpath(@__DIR__, "tricks_common.jl"))
apply_theme!()
const OUT = output_dir(@__DIR__)


function build_secular_determinant()
    @syms x::real E::real
    ell = SymPy.Sym(2)
    a = [SymPy.symbols("a_$(k)") for k in 1:5]
    u = (1 - x^2) * sum(a[k] * x^(2(k-1)) for k in 1:5)
    u_x  = diff(u, x)
    u_xx = diff(u, x, 2)

    R = (1 - x^2)^4 * ((1 - x^2) * u_xx - 3 * x * u_x) / ell^2 +
        ((1 - x^2)^2 * E - ell^4 * x^4) * u

    eqs = [SymPy.integrate(x^(2j) * R, (x, -1, 1)) for j in 0:4]

    M = SymPy.Sym[simplify(eqs[i].coeff(a[j])) for i in 1:5, j in 1:5]
    Mmat = SymPy.Sym.(reshape(M, 5, 5))
    D_E = simplify(symbolic_det(Mmat))
    poly_E = sympy.Poly(D_E, E)
    return poly_E, x, E
end


# Hand-rolled cofactor-expansion determinant for symbolic matrices.
function symbolic_det(A::AbstractMatrix{<:SymPy.Sym})
    n = size(A, 1)
    n == 1 && return A[1, 1]
    s = SymPy.Sym(0)
    for j in 1:n
        sub = A[setdiff(1:n, 1), setdiff(1:n, j)]
        s += (-1)^(j + 1) * A[1, j] * symbolic_det(sub)
    end
    return s
end


function main_run(dump_path)
    poly_E, x_sym, E_sym = build_secular_determinant()
    coeffs_E = poly_E.all_coeffs()       # high-degree first
    println("Symbolic secular determinant (high degree -> low):")
    for (k, c) in enumerate(coeffs_E)
        println("   E^$(length(coeffs_E) - k): $(simplify(c))")
    end
    norm_const = Float64(coeffs_E[end])
    println("\nBoyd Eq 20.16 form  D(E) / D(0):")
    for (k, c) in enumerate(reverse(coeffs_E))
        println("   E^$(k - 1): $(Float64(c) / norm_const)")
    end

    coeffs_num = Float64.(coeffs_E)
    poly = Polynomial(reverse(coeffs_num))   # Polynomials.jl: low to high
    rts = Polynomials.roots(poly)
    rts_real = sort([real(r) for r in rts if abs(imag(r)) < 1e-6])

    bender_orszag = [1.060362, 7.45570, 16.2618, 26.528, 37.92]
    println("\nEigenvalues from secular determinant vs Bender-Orszag:")
    for (r, ref) in zip(rts_real, bender_orszag)
        rel = (r - ref) / ref
        println("   E_num = $(round(r; digits = 4))    E_exact = $ref    rel err = $(round(rel; sigdigits=3))")
    end

    fig = Figure(size = (1150, 440))

    E_lo = 0.0; E_hi = 50.0
    Es = collect(range(E_lo, E_hi; length = 2000))
    Ds = poly.(Es)
    log_absD = log10.(abs.(Ds) .+ 1e-30)
    rug_y = minimum(log_absD) - 0.6

    ax1 = Axis(fig[1, 1]; xlabel = "eigenvalue E",
               ylabel = "log10 |D(E)|",
               title = "secular determinant zeros on the lower window E in [0, 50]",
               limits = ((E_lo, E_hi), (rug_y - 2.2, maximum(log_absD) + 0.5)))
    lines!(ax1, Es, log_absD; color = NAVY, linewidth = 1.0,
           label = "log10 |D(E)|")

    in_window(v) = E_lo <= v <= E_hi
    refs_in   = filter(in_window, bender_orszag)
    roots_in  = filter(in_window, rts_real)
    roots_out = filter(v -> !in_window(v), rts_real)

    for ref in refs_in
        vlines!(ax1, [ref]; color = TEAL, linewidth = 0.6, alpha = 0.35,
                linestyle = :dot)
    end
    scatter!(ax1, refs_in, fill(rug_y, length(refs_in));
             color = :white, strokecolor = TEAL, strokewidth = 1.2,
             markersize = 10,
             label = "Bender-Orszag E_n (reference)")
    scatter!(ax1, roots_in, fill(rug_y, length(roots_in));
             color = CORAL, marker = :xcross, markersize = 14,
             label = "numerical roots of D")
    if !isempty(roots_out)
        labs = ["E_6", "E_8"]
        parts = String[]
        for (k, r) in enumerate(roots_out[1:min(end, 2)])
            push!(parts, "$(labs[k]) ≈ $(round(r; digits=0))")
        end
        annot_text = join(parts, ", ") * " (numerical mirages, see right panel)"
        text!(ax1, E_hi - 28, rug_y - 1.6;
              text = annot_text,
              color = CORAL, fontsize = 8)
    end
    axislegend(ax1, position = :lt, labelsize = 9)

    ns = [0, 2, 4, 6, 8]
    rels = abs.([(r - ref) / ref for (r, ref) in zip(rts_real, bender_orszag)])
    ax2 = Axis(fig[1, 2]; xlabel = "physical mode index n", ylabel = "relative error",
               yscale = log10,
               title = "trust the lower spectrum: bottom 3 good, top 2 unusable",
               limits = ((-0.5, 8.5), (1e-4, 1e2)))
    scatterlines!(ax2, ns, rels; color = NAVY, markercolor = :white,
        strokecolor = NAVY, strokewidth = 1.0, markersize = 8, linewidth = 1.0,
        label = "|E_n^sym - E_n| / E_n")
    hlines!(ax2, [0.01]; color = TEAL, linewidth = 0.4, alpha = 0.5,
            label = "1% threshold")
    axislegend(ax2, position = :lt, labelsize = 9)

    save(joinpath(OUT, "symbolic_quartic_oscillator.pdf"), fig)
    save(joinpath(OUT, "symbolic_quartic_oscillator.png"), fig, px_per_unit = 4)
    println("  figure: ", joinpath(OUT, "symbolic_quartic_oscillator.pdf"))

    if dump_path !== nothing
        write_json(dump_path, Dict("ell" => 2.0,
            "secular_coeffs" => coeffs_num, "roots" => rts_real,
            "bender_orszag" => bender_orszag, "rel_err" => rels))
    end
end


function parse_args()
    dp = nothing; i = 1
    while i <= length(ARGS)
        if ARGS[i] == "--dump" && i < length(ARGS)
            dp = ARGS[i+1]; i += 2
        else; i += 1; end
    end
    return dp
end


if abspath(PROGRAM_FILE) == @__FILE__
    main_run(parse_args())
end

File: codes/python/ch21/symbolic_quartic_oscillator.py
#!/usr/bin/env python3
"""
symbolic_quartic_oscillator.py

Chapter 21: Special Tricks for the Spectral Researcher
Computational Etude 21.10: the quartic oscillator on an infinite interval,
                           solved symbolically.

The eigenvalue problem
        u_yy + (E - y^4) u = 0,    |u| -> 0  as |y| -> infty,
is the classic 'quartic oscillator' of quantum mechanics.  We attack it
in three stages:

1. Map the line to (-1, 1) via the rational-Chebyshev coordinate
        y = ell x / sqrt(1 - x^2).
   In x-space the equation becomes (Boyd Eq 20.13)
        (1-x^2)^4 {(1-x^2) u_xx - 3 x u_x} / ell^2
        + {(1-x^2)^2 E - ell^4 x^4} u  =  0.

2. Build in the boundary condition u(+/-1) = 0 by writing
        u(x) = (1 - x^2) (a_1 + a_2 x^2 + a_3 x^4 + a_4 x^6 + a_5 x^8).
   This is 'basis recombination' (Boyd Sec 20.2): a five-term ansatz
   that automatically satisfies the BC.

3. Demand orthogonality of the residual against the test functions
        {1, x^2, x^4, x^6, x^8}
   on x in [-1, 1].  This produces a 5x5 LINEAR system whose entries
   depend linearly on E.  The vanishing of its determinant is a
   degree-5 polynomial in E -- the 'secular determinant' D(E).

We compare the resulting D(E) to Boyd Eq 20.16 (after normalising the
constant term to 1) and the resulting eigenvalues to the high-precision
reference of Bender and Orszag (Adv Math Methods, 1978, p 523):

        E_0 = 1.060362,  E_2 = 7.45570,  E_4 = 16.2618,
        E_6 = 26.528,    E_8 = 37.92.

The lower three eigenvalues are accurate to 1-2 decimals; the upper two
are spectacularly inaccurate.  This is Boyd's enduring symbolic lesson:
trust the LOWER spectrum first.

Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
Part of "Computational Etudes: A Spectral Approach"
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import sympy as sp

SCRIPT_DIR = Path(__file__).resolve().parent
sys.path.insert(0, str(SCRIPT_DIR))
from tricks_common import setup_matplotlib, output_dir_for, save_fig, dump_json
from tricks_common import NAVY, CORAL, TEAL


ELL = sp.Integer(2)


def build_secular_determinant():
    x, E = sp.symbols("x E", real=True)
    ell = ELL
    a = sp.symbols("a1:6")  # (a_1, a_2, a_3, a_4, a_5)

    u = (1 - x ** 2) * sum(a[k] * x ** (2 * k) for k in range(5))
    u_x = sp.diff(u, x)
    u_xx = sp.diff(u, x, 2)

    # Boyd Eq 20.13 (multiplied through by (1-x^2)^2 to clear the y^4 pole).
    R = (1 - x ** 2) ** 4 * ((1 - x ** 2) * u_xx - 3 * x * u_x) / ell ** 2 \
        + ((1 - x ** 2) ** 2 * E - ell ** 4 * x ** 4) * u

    # Galerkin moments with test functions x^{2j}, j = 0, 1, 2, 3, 4.
    eqs = [sp.integrate(x ** (2 * j) * R, (x, -1, 1)) for j in range(5)]

    # Build the 5x5 matrix in the unknowns (a_1, ..., a_5).
    M = sp.Matrix([[sp.simplify(eqs[i].coeff(a[j])) for j in range(5)] for i in range(5)])
    D_expr = sp.Poly(sp.simplify(M.det()), E)
    return D_expr, M, x, E


def main():
    p = argparse.ArgumentParser(description=__doc__.splitlines()[1])
    p.add_argument("--dump", type=str, default=None)
    args = p.parse_args()

    setup_matplotlib()
    out = output_dir_for(SCRIPT_DIR)

    D_poly, M, x, E = build_secular_determinant()
    coeffs_E = D_poly.all_coeffs()                      # high-degree first
    print("Symbolic secular determinant (high degree -> low):")
    for k, c in enumerate(coeffs_E):
        print(f"   E^{len(coeffs_E)-1-k}: {sp.nsimplify(c)} = {float(c):.6e}")
    norm_const = float(coeffs_E[-1])
    print("\nBoyd Eq 20.16 form  D(E) / D(0) = 1 + c_1 E + c_2 E^2 + ...:")
    for k, c in enumerate(coeffs_E[::-1]):
        print(f"   E^{k}: {float(c) / norm_const:.6e}")

    # Find numerical roots of the polynomial in E.
    coeffs_num = [float(c) for c in coeffs_E]
    roots = np.roots(coeffs_num)
    roots_real = sorted([float(np.real(r)) for r in roots if abs(np.imag(r)) < 1e-6])

    bender_orszag = [1.060362, 7.45570, 16.2618, 26.528, 37.92]   # n = 0, 2, 4, 6, 8
    print("\nEigenvalues from secular determinant vs Bender-Orszag (n = 0, 2, 4, 6, 8):")
    for r, ref in zip(roots_real, bender_orszag):
        rel = (r - ref) / ref
        print(f"   E_num = {r:11.4f}    E_exact = {ref:9.4f}    rel err = {rel:+.2e}")

    # ---- Figure ------------------------------------------------------
    fig = plt.figure(figsize=(11.5, 4.4))
    gs = fig.add_gridspec(1, 2, wspace=0.30)

    # Panel A: log10 |D(E)| on the relevant window [0, 50] -- zeros
    # appear as deep V-shaped dips, sharply localised, easy to read off.
    # The five Bender-Orszag references are placed as a "rug" of teal
    # circles on the x-axis; coral X markers locate the numerical roots
    # the symbolic determinant actually has in this window.  The lone
    # rogue root (the wholly-worthless E_8 ~ 720) is annotated with
    # an arrow rather than allowed to stretch the axis.
    ax = fig.add_subplot(gs[0, 0])
    E_window = (0.0, 50.0)
    Es = np.linspace(E_window[0], E_window[1], 2000)
    Ds = np.polyval(coeffs_num, Es)
    log_absD = np.log10(np.abs(Ds) + 1e-30)
    ax.plot(Es, log_absD, color=NAVY, linewidth=1.0, label=r"$\log_{10}|D(E)|$")

    # Roots and references inside the window
    in_window = lambda v: E_window[0] <= v <= E_window[1]
    roots_in = [r for r in roots_real if in_window(r)]
    refs_in  = [r for r in bender_orszag if in_window(r)]
    roots_out = [r for r in roots_real if not in_window(r)]

    # Rug strip just below the curve range
    rug_y = log_absD.min() - 0.6
    ax.scatter(refs_in, [rug_y] * len(refs_in), s=70, marker="o",
               facecolors="none", edgecolors=TEAL, linewidth=1.2,
               label=r"Bender-Orszag $E_n$ (reference)")
    ax.scatter(roots_in, [rug_y] * len(roots_in), s=80, marker="x",
               color=CORAL, linewidth=1.5,
               label="numerical roots of $D$")
    # vertical guide lines at references
    for ref in refs_in:
        ax.axvline(ref, color=TEAL, linewidth=0.6, alpha=0.35, linestyle=":")

    # Annotate the out-of-window rogue roots (the failure modes).
    # There are two: E_6 ~ 64 (way past its reference 26.5) and E_8 ~ 716
    # (way past its reference 37.9). Both are mirages; both deserve mention.
    if roots_out:
        labels = [r"$E_6$", r"$E_8$"]
        text_parts = [fr"{labels[i]} $\approx$ {r:.0f}"
                      for i, r in enumerate(roots_out[:2])]
        annot_text = ", ".join(text_parts) + " (numerical mirages, see right panel)"
        ax.annotate(annot_text,
                    xy=(E_window[1] - 0.5, rug_y),
                    xytext=(E_window[1] - 28, rug_y - 1.6),
                    color=CORAL, fontsize=8,
                    arrowprops=dict(arrowstyle="->", color=CORAL,
                                    lw=0.8, shrinkA=2, shrinkB=2))

    ax.set_xlabel(r"eigenvalue $E$")
    ax.set_ylabel(r"$\log_{10}|D(E)|$")
    ax.set_title(r"secular determinant zeros on the lower window $E\in[0,50]$",
                 fontsize=10)
    ax.set_xlim(*E_window)
    ax.set_ylim(rug_y - 2.2, log_absD.max() + 0.5)
    ax.legend(loc="upper left", fontsize=9)
    ax.grid(True, alpha=0.25, linewidth=0.4)

    # Panel B: relative error vs mode index
    ax = fig.add_subplot(gs[0, 1])
    ns = [0, 2, 4, 6, 8]
    rels = np.abs([(r - ref) / ref for r, ref in zip(roots_real, bender_orszag)])
    ax.semilogy(ns, rels, "o-", color=NAVY, markerfacecolor="white",
                markersize=8, linewidth=1.0,
                label=r"$|E_n^{\mathrm{sym}} - E_n| / E_n$")
    ax.axhline(0.01, color=TEAL, linewidth=0.4, alpha=0.5,
               label="1% threshold (lower spectrum target)")
    ax.set_xlabel(r"physical mode index $n$ (even modes only)")
    ax.set_ylabel("relative error")
    ax.set_title("trust the lower spectrum: bottom 3 modes good, top 2 unusable",
                 fontsize=10)
    ax.legend(loc="upper left", fontsize=9)
    ax.grid(True, alpha=0.25, linewidth=0.4)
    ax.set_ylim(1e-4, 1e2)

    save_fig(fig, out, "symbolic_quartic_oscillator")
    plt.close(fig)

    print(f"  figure: {out / 'symbolic_quartic_oscillator.pdf'}")

    dump_json(args.dump, {
        "ell": float(ELL),
        "secular_coeffs": [float(c) for c in coeffs_E],
        "roots": roots_real,
        "bender_orszag": bender_orszag,
        "rel_err": rels.tolist(),
    })


if __name__ == "__main__":
    main()
